daemons/fenced/fenced_history.c

```diff
@@ -363,25 +382,13 @@ stonith_merge_in_history_list(GHashTable *history)
             g_hash_table_lookup(stonith_remote_op_list, op->id);
 
         if (stored_op) {
-            if (stored_op->state != st_failed
-                && stored_op->state != st_done
-                && (op->state == st_failed || op->state == st_done)) {
-                    crm_debug("Updating outdated pending operation %.8s "
-                              "(state=%s) according to the one (state=%s) from "
-                              "remote peer history",
-                              op->id, stonith_op_state_str(stored_op->state),
-                              stonith_op_state_str(op->state));
-
-            } else {
-                continue; // Skip existent
-            }
+            continue; // Skip existent
         }
 
         updated = TRUE;
```

Comment:
maybe we should move that after the block that is failing the pending.

Reply:
What if we toss the current merge function, rename stonith_local_history_diff to stonith_local_history_diff_and_merge and make stonith_local_history a simple wrapper for it that doesn't have then received list in the interface.

Reply:
Yes, we can make it clearer and easier for callers. Probably we should anyway keep the merging function for stonith_local_history_diff_and_merge() 's  own use in order to prevent it from growing too big. How about something this?

Reply:
Had the impression that the merge would degenerate to something that is
- not large enough anymore to be necessarily split into an extra function for a single use
- doesn't have a functionality (anymore) that makes it easy to give it a name that expresses what it is doing
but if it looks too long do whatever it needs ;-)
Maybe 'append' instead to express that the 'merge' logic is actually gone ...

Reply:
Aah yeah ... one more thing I just realized:
The updated-logic has to cover of course what we've done in our first loop.
But I would still do - for the reasons given in my comment - just a single trim.
If with spawned out merge/append/whatever maybe passing a pointer to the 
updated-flag modified in the first loop to modify in case.

Reply:
Forget the above it is garbage. Whenever we remove something from our current list we have an update for it.

Reply:
Unless we are doing a probably more efficient g_hash_table_iter_replace in the
current table and an immediate g_hash_table_steal(_extended - probably needed 
as we need to free the stolen key) in the remote-list.

Reply:
crap above suggestion would still not work.
easiest would probably be simply switching id between op and remote_op before the
g_hash_table_steal and subsequent
g_hash_table_iter_replace (frees the old op with the remote_op-id).

Reply:
Not sure if it's a little hacky. I don't mind though :-)

Reply:
Tricking it to free what we want is a bit hacky but sacrificing beauty to speed is fine ;-)
For the length of the function I think it is important to keep functions short that have
complex execution logic but if we just have linear executed code ...